Distribution Specs (v3) - Builder Distribution Spec (#271)
buildpacks/spec: distribution.md

/cnb/buildpacks/<buildpack ID>/<buildpack version>/
```

> If the buildpack ID contains a `/`, it must be replaced with `_` in the directory name.

Reviewer: ```suggestion
If the buildpack ID contains a `/`, it MUST be replaced with `_` in the directory name.
```

    * contain an array of deprecated APIs:
      * should only contain `0.x` or major versions
      * should only contain APIs that correspond to a spec release

### Builder

The following specifies the artifact format for a buildpacks builder.

A Builder is an OCI Image that provides a distributable build environment.

A platform supporting the builder SHOULD allow users to configure the build environment with a provided builder.

#### General Requirements

The Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and [build-time](#build-image) environment and MAY contain [buildpacks](#buildpackage). Platforms SHOULD use builders to ease the build process. 

#### Environment Variables

A Builder MUST be an extension of a build-image, and MUST retain all the specified environment variables set on the original build image, as specified in the Build Image specifications.

#### Labels

> A Builder MUST be an extension of a build-image, and MUST retain all the specified Labels set on the original build image, as specified in the Build Image specifications.

Reviewer 2: ```suggestion
A Builder MUST be an extension of a Build Image, and MUST retain all the specified Labels set on the original build image, as specified in the Build Image specifications.
```
let's avoid hyphenating these, to match the other sections

Author: Done!

> A Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

Author: @jkutner what I meant by that line is that the Builder image (which MUST contain the lifecycle implementation) must also have the labels specified on the lifecycle image. So if I have to distribute a Builder Image, it must have all the labels set on lifecycle because it contains the lifecycle implementation.

> A Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

Reviewer 3: @jimil749 are those labels spec'ed somewhere?

> A Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

Author: Yep, in the lifecycle distribution specification. https://github.com/buildpacks/spec/pull/264

> A Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

Author: The goal is to reference the buildpack components i.e. build image, run image, buildpacks and lifecycle from the builder image.

> A Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

Reviewer 4: @jkutner do you want any changes for this?

> A Builder image MAY contain [buildpacks](#buildpackage), and MAY retain all the specified Labels set on the original buildpackage, as specified in the buildpackage specifications.

Reviewer 3: Should we say where the buildpacks live on the filesystem?

Author: /cc @jromero

Reviewer 5: I believe that is already covered [by this](https://github.com/buildpacks/spec/blob/cbfaca6e52af1c89b32b0350bffedf9f4c1b4175/distribution.md?plain=1#L43-L49):

![image](https://user-images.githubusercontent.com/475559/144277414-39a0e7bf-ead5-4d5a-937e-75e9e369cbd1.png)

Maybe we can add some additional clarification but I think we should be doing so in the "Buildpackage" section and not in this section. Thoughts?

Reviewer 5: This part seems important from the [previously unreleased version of the builder spec](https://github.com/buildpacks/spec/blob/extensions/builder/0.1/extensions/builder.md#filesystem). I would still propose that this be added to the "buildpackages" section.

![image](https://user-images.githubusercontent.com/475559/144280287-2603c416-d8f8-4570-bcef-af4d17a4c56e.png)

Author: I've mentioned the location of the buildpack blob in the buildpackage section as in the link you mentioned above, I'll refine the section.

Author: @jromero would adding `CNB_APP_DIR` and `CNB_LAYERS_DIR` in buildpackage section make sense? Because essentially it is referring to the filesystem of the Builder image. We can have a sub-section here, under the builder section and this information there. I left out the path of the buildpack blob because that information is already present in the buildpack specs [here](https://github.com/buildpacks/spec/blob/cbfaca6e52af1c89b32b0350bffedf9f4c1b4175/distribution.md?plain=1#L43-L49).

Author: @jromero would adding `CNB_APP_DIR` and `CNB_LAYERS_DIR` in buildpackage section make sense? Because essentially it is referring to the filesystem of the Builder image. We can have a sub-section here, under the builder section and this information there. I left out the path of the buildpack blob because that information is already present in the buildpack specs [here](https://github.com/buildpacks/spec/blob/cbfaca6e52af1c89b32b0350bffedf9f4c1b4175/distribution.md?plain=1#L43-L49).

Reviewer 5: I captured too much in my screenshot. I'm sorry. You are right those don't make sense in the Buildpackages section. We will need them in this PR though. 

I believe the [environment variables](https://github.com/buildpacks/spec/blob/extensions/builder/0.1/extensions/builder.md#environment-variables) defined here need to be included. The exception is `CNB_BUILDPACKS_DIR`. That brings about an issue if we want to reuse layers from Buildpackages. I would propose that we remove this environment variable. cc: @buildpacks/core-team 

---

The part that is missing from my screenshot is:

> If the buildpack ID contains a `/`, it MUST be replaced with `_` in the directory name.

That's very critical.

Author: I've added that to the buildpackage section.

The following labels MUST be set in the builder environment (through the image config's `Labels` field):

| Label             | Description
| --------          | --------
| `io.buildpacks.builder.metadata`     | A JSON object representing Builder metadata
| `io.buildpacks.buildpack.order`      | A JSON object representing the order of the buildpacks stored on the builder

`io.buildpacks.builder.metadata` (JSON)

```json
{
  "description": "<description>",
  "buildpacks": [
    {
      "id": "<buildpack ID>",
      "version": "<buildpack version>",
      "homepage": "<buildpack homepage>",
    }
  ],
  "createdBy": {
    "name": "<tool name>",
    "version": "<tool version>",
  }
}
```

The `createdBy` metadata is meant to contain the name and version of the tool that created the builder.

`io.buildpacks.buildpack.order` (JSON)

```json
[
  {
    "group":
      [
        {
          "id": "<buildpack ID>",
          "version": "<buildpack version>",
          "optional": false
        }
      ]
  }
]

> ```

Reviewer 5: Based on the previous unreleased builder spec, we did have a "Filesystem" section. Looking at it in depth I think most things are covered by _inheriting_ layers from the other components (ie. lifecycle binaries from lifecycle image, buildpack layers from buildpackages, etc).

The one file that I want to discuss on whether it is necessary is `order.toml`. Builders currently have that file on the filesystem and should match the `io.buildpacks.buildpack.order` label. In many cases, platforms override or create a new version of this file before providing it to the lifecycle. Do we need to keep a filesystem format of the order?

In a theoretical world where it didn't exist, this is how existing known platforms could handle it.

 * pack - can easily add that file as a layer (we do this already when buildpacks are added).
 * kpack - already [creates an order.toml dynamically](https://github.com/pivotal/kpack/blob/ffd8cd1caf1e852df4e714d1cb8b9f4e7dd89e1b/pkg/cnb/builder_builder.go#L250-L287).
 * tekton - as part of "prepare" could inspect the image config and generate the file.
 
 Personally, I'm not seeing the value or requiring it in both places.

Reviewer 5: @sclevine / @jkutner / @natalieparellano / @jabrown85: Do you all have an option here?

Reviewer 5: `pack` would do the same as Tekton and create an `order.toml` based on the image label.

Reviewer 6: I think we should leave `order.toml` required. It doesn't seem great that a simple platform like tekton would have to read the label instead of being able to rely on it being in the builder itself. It means platforms MUST be able to read labels prior to executing builds on them. If the platform allows customers to bring their own builder, it is much easier to assume the the builder knows how to resolve the order itself and the platform would have less work to do.

it also makes testing a builder harder and more magic IMO. Today you can do something like this to test out the builder image without having to use `pack` or anything else special.

```
docker run -it -v $(pwd):/workspace --entrypoint /cnb/lifecycle/creator heroku/buildpacks:20 myimg
```

That wouldn't be possible if the file is removed.

Reviewer 5: I concede to your argument @jabrown85. I do feel it's not optimal but can see your points.

What do we think about the format? As you define it @jabrown85, it makes it seem like the format should be part of the distribution spec and not the `platform` spec. Today, it defers out to a platform API defined [format][format] when it's embedded into a distributed package. I don't see how that would work in your scenarios. I would argue that the format should be defined by the distribution spec. The platform could still create a new one if need be but it should do so while considering the distribution version.

[format]: https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml

Reviewer 5: I'll stand down from my contentious point above and defer it to a separate RFC if need be. 

@jimil749 do you think we can add the notion of `order.toml`, its location `/cnb/order.toml`, and reference to its [format][format] in this spec?

[format]: https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml

Author: @jromero, I will get to it asap

Author: @jromero sorry for the delay, but I've added a `Filesystem` section wherein I've mentioned that `order.toml` MUST be present in the builder fs and referenced the platform specification for the format.

A platform supporting the builder SHOULD allow users to configure the build environment with a provided builder.

#### General Requirements

The Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and [build-time](#build-image) environment and MAY contain [buildpacks](#buildpackage). Platforms SHOULD use builders to ease the build process. 

#### Environment Variables

A Builder MUST be an extension of a Build Image, and MUST retain all the specified environment variables set on the original build image, as specified in the Build Image specifications.

The following environment variables MUST be set on the builder (through the image config's `Env` field):

| Env Variable             | Description                                                                       | Default
| --------                 | --------                                                                          |-----------
| `CNB_APP_DIR`            | Application directory of the build environment                                    | `/workspace`
| `CNB_LAYERS_DIR`         | The directory to create and store `layers` in the build environment               | `/layers`

> | `CNB_PLATFORM_DIR`       | The directory to create and store platform focused files in the build environment | `/platform`

Reviewer 5: I'm curious why there are "defaults" if these are required environment variables.

EDIT: Seems like @ekcasey had the same thoughts in [this comment](https://github.com/buildpacks/spec/pull/193/files#r635266996) on the original source.

I propose we change it to MAY instead of MUST. If we make that change, it could be merged with the table below.

Author: Ahh yes, we can change it to be MUST.

Where,

`<Target ID>` is an identifier specified on the runtime image that MAY be used to apply target-specific logic.
### Builder

The following specifies the artifact format for a buildpacks builder.

A Builder is an OCI Image that provides a distributable build environment.

> A platform supporting the builder SHOULD allow users to configure the build environment with a provided builder.

Reviewer: ```suggestion
A platform supporting builders SHOULD allow users to configure the build environment for a provided builder.
```
(maybe... in my head this makes more sense!)

#### General Requirements

> The Builder image MUST contain an implementation of the [lifecycle](#lifecycle), and [build-time](#build-image) environment and MAY contain [buildpacks](#buildpackage). Platforms SHOULD use builders to ease the build process. 

Reviewer: ```suggestion
The Builder image MUST contain an implementation of the [lifecycle](#lifecycle) and a [build-time](#build-image) environment, and MAY contain [buildpacks](#buildpackage). Platforms SHOULD use builders to ease the build process. 
```

Author: Done

Where,

`<Target ID>` is an identifier specified on the runtime image that MAY be used to apply target-specific logic.

### Builder

> The following specifies the artifact format for a buildpacks builder.

Reviewer: nit: sometimes we use "builder" and sometimes "Builder". We should probably pick one and run with it. I vote for "builder".

Author: Sure, I will make the changes

A builder is an OCI Image that provides a distributable build environment.

A platform supporting builders SHOULD allow users to configure the build environment for a provided builder.

#### General Requirements

The builder image MUST contain an implementation of the [lifecycle](#lifecycle), and a [build-time](#build-image) environment, and MAY contain [buildpacks](#buildpackage). Platforms SHOULD use builders to ease the build process.

#### Filesystem

A builder MUST have the following directories/files:

> * `/cnb/order.toml` &rarr; As defined in the [platform specification](https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml)

Reviewer 7: I suggest that we remove this requirement entirely given that we haven't thought about how to handle the case where we change or remove this file between platform APIs. The platform can always pass in a valid `order.toml` for it's preferred platform API, created from the label if it wants to use the builder-provided default.

Reviewer 5: I was in the [same boat](https://github.com/buildpacks/spec/pull/271/files#r760387860) for this and other reasons. I eventually caved in to minimize the turbulence. Happy to revisit but would like to take @jabrown85's [points](https://github.com/buildpacks/spec/pull/271/files#r767823046) into consideration.

Reviewer 5: I'd love to defer this to a separate talking point, if you feel strongly we can even block the release of the distribution spec on it.

Reviewer 5: Punted to https://github.com/buildpacks/spec/issues/291

#### Environment Variables

A builder MUST be an extension of a Build Image, and MUST retain all the specified environment variables set on the original build image, as specified in the Build Image specifications.

The following environment variables MAY be set on the builder (through the image config's `Env` field):

| Env Variable             | Description                                                                       | Default
| --------                 | --------                                                                          |-----------

> | `CNB_APP_DIR`            | Application directory of the build environment                                    | `/workspace`

Reviewer 7: For this and other directory env vars, should we say something to effect that these directories should exist an be writable by the build image user? In my mind that is the only reason to require these env vars in the builder spec.

Reviewer 5: Resolved in 35e5cce6acb83cfa018f61dc0bc90019375b8702

| `CNB_LAYERS_DIR`         | The directory to create and store `layers` in the build environment               | `/layers`
| `CNB_PLATFORM_DIR`       | The directory to create and store platform focused files in the build environment | `/platform`

> | `SERVICE_BINDING_ROOT`   | The directory where services are bound                                            | -

Reviewer 7: I don't think including defaults make sense here. The defaults are a piece of lifecycle behavior, specified in the platform API and don't belong in this specification

Reviewer 5: Resolved in 35e5cce6acb83cfa018f61dc0bc90019375b8702

#### Labels

A builder MUST be an extension of a Build Image, and MUST retain all the specified Labels set on the original build image, as specified in the Build Image specifications.

A builder image MUST contain an implementation of the [lifecycle](#lifecycle), and MUST retain all the specified Labels set on the original Lifecycle image, as specified in the lifecycle distribution specifications.

> A builder image MAY contain [buildpacks](#buildpackage), and MAY retain all the specified Labels set on the original buildpackage, as specified in the buildpackage specifications.

Reviewer 7: I think this section needs some adjutment. Multiple buildpackages will have conflicting (non-compatible) `io.buildpacks.buildpackage.metadata` labels and any `io.buildpacks.buildpackage.layers` labels will need to be merged.

Reviewer 5: I can think of two options:

1. Make the labels unique per buildpack (ie. `io.buildpacks.buildpackage.<id>.<version>.metadata` and `io.buildpacks.buildpackage.<id>.<version>.layers`).
    - Drawbacks: 
        - Predictability and discoverability. Usage would require a more complex glob search in all labels.
        - Duplication. Potential duplication of data when a buildpack is present in multiple meta-buildpacks.
2. `io.buildpacks.buildpackage.metadata` and `io.buildpacks.buildpackage.layers` must be unset and a new label `io.buildpacks.builder.layers` must be created merging `io.buildpacks.buildpackage.layers`.
    - Drawbacks:
        - Additional processing of labels.
        - Can't "duck-type" a builder as a buildpackage. (Maybe that's a good thing).

I'm personally leaning towards option 2. What are your thoughts? Anything I'm overlooking?

Reviewer 7: I think option 3 is to stick with the existing `io.buildpacks.buildpack.layers` merged label. Is there a reason we don't like that option?

Reviewer 5: Did you mean `io.buildpacks.buildpackages.layers` or are you proposing renaming what is already in this branch of the spec?

That aside, I see what you mean. Option 3 is a merge and update of the existing label. Your description doesn't  account for the _other_ buildpackage label `...metadata`. I still think that should be unset.

Personally, I don't have a preference between option 2 or 3+. 

If I were to summarize, what do you think of:

1. Rename `io.buildpacks.builpackages.layers` => `io.buildpacks.builpack.layers`
2. Rename `io.buildpacks.builpackages.metadata` => `io.buildpacks.builpack.metadata`
3. Merge all `io.buildpacks.buildpack.layers` of any buildpackages into `io.buildpacks.builpack.layers` when a builder is built
4. `io.buildpacks.buildpack.metadata` should not be present on a builder image.

_The rename would also align better with `io.buildpacks.buildpack.order` below._

Reviewer 7: I like option 3+

While we would need to rename `io.buildpacks.buildpackages.layers` in this branch of the spec with conveniently **would not** need to rename the label (`io.buildpacks.builpack.layers`) that we are currently applying to buildpacks and builders

Reviewer 5: Resolved in 35e5cce6acb83cfa018f61dc0bc90019375b8702

The following labels MUST be set in the builder environment (through the image config's `Labels` field):

| Label             | Description
| --------          | --------
| `io.buildpacks.builder.metadata`     | A JSON object representing builder metadata
| `io.buildpacks.buildpack.order`      | A JSON object representing the order of the buildpacks stored on the builder

`io.buildpacks.builder.metadata` (JSON)

```json
{
  "description": "<description>",
  "buildpacks": [
    {
      "id": "<buildpack ID>",
      "version": "<buildpack version>",
      "homepage": "<buildpack homepage>",
    }

>   ],

Reviewer 7: Is this duplicative of `io.builpacks.buildpackage.layers`. Should we remove this section but require `io.builpacks.buildpackage.layers` on builders?

Reviewer 5: If we go with option 2 of [this comment](https://github.com/buildpacks/spec/pull/271/files#r791819949) then this may also be cleared up.

Reviewer 5: Resolved in 35e5cce6acb83cfa018f61dc0bc90019375b8702
